modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TestTcpDiscoverySpi.java

```diff
@@ -44,4 +54,24 @@
     @Override public void simulateNodeFailure() {
         super.simulateNodeFailure();
     }
+
+    /** {@inheritDoc} */
+    @SuppressWarnings("ZeroLengthArrayAllocation")
+    @Override public void setListener(@Nullable DiscoverySpiListener lsnr) {
+        super.setListener(lsnr ==  null || discoHooks.isEmpty() ?
```

Comment:
Is the following flow possible? If no, why?

1. addDiscoveryHook
2. setListener
3. addDiscoveryHook <- this hook will be ignored.

Reply:
It seems to me that we can just pass discoHooks into the listener.
Therefore list always be up to date inside the listener.

Reply:
addDiscoveryHook method asserts that it must be called only before TcpDiscoverySpi start. So it's impossible to use addDiscoveryHook after listener is set.

Reply:
OK then.

Reply:
> It seems to me that we can just pass discoHooks into the listener.
> Therefore list always be up to date inside the listener.

It's not clear for me in what cases hooks are needed to be modified after SPI start. I think that the current approach is more robust.